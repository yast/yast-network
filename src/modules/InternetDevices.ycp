/****************************************************************************
 *
 * Copyright (c) 2012 Novell, Inc.
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail,
 * you may find current contact information at www.novell.com
 *
 ****************************************************************************/
/**
 * File:	modules/InternetDevices.ycp
 * Package:	Network configuration
 * Summary:	Internet connection and YOU during the installation
 * Authors:	Michal Svec <msvec@suse.cz>
 *		Arvin Schnell <arvin@suse.de>
 *
 */

{

module "InternetDevices";

import "Internet";
import "NetworkInterfaces";

/**
 * Reset values.
 */
define void Reset() {
    Internet::device = "";
    Internet::type = "";
    Internet::logfile = "";
    Internet::provider = "";
    Internet::password = "";
    Internet::demand = false;
    Internet::askpassword = nil;
    Internet::capi_adsl = false;
    Internet::capi_isdn = false;
}

/**
 *  Set device from argument as default network device 
 */
global define void SetDevice(string dev){
    Internet::device=dev;
    NetworkInterfaces::Select(Internet::device);
    Internet::type = NetworkInterfaces::FastestType(Internet::device);
    Internet::provider = NetworkInterfaces::Current["PROVIDER"]:"";

    if (Internet::provider != "") {
        import "Provider";

        Provider::Read();
        Provider::Select(Internet::provider);

        Internet::demand = Provider::Current["DEMAND"]:"no" == "yes";
        Internet::password = Provider::Current["PASSWORD"]:"";
        Internet::askpassword = Provider::Current["ASKPASSWORD"]:"no" == "yes";
        Internet::capi_adsl = Provider::Current["PPPMODE"]:"pppoe" == "capi-adsl";
        Internet::capi_isdn = Provider::Current["PPPMODE"]:"ippp" == "capi-isdn";
    }

    /* Set logfile */
    if (Internet::type == "dsl" || Internet::type == "modem" || (Internet::type == "isdn" && Internet::capi_isdn))
        Internet::logfile = "/var/log/smpppd/ifcfg-" + Internet::device + ".log";

}

/**
 * Find the fastest connection to the Internet
 * @return true if a "good" connection was found
 */
global define boolean FindFastest() {

    Reset();

    NetworkInterfaces::Read();
    Internet::device = NetworkInterfaces::Fastest();

    if (Internet::device == "")
    {
	Internet::device = Internet::GetDevices ()[0]:"";
    }

    y2milestone("fastest=%1", Internet::device);

    /* No fallback since there are devices that must not be tested (e.g. lo) */
    if (Internet::device == "") return false;

    SetDevice (Internet::device);
    return true;
}

/* EOF */
}
