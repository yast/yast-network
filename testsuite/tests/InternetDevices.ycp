/****************************************************************************
 *
 * Copyright (c) 2012 Novell, Inc.
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail,
 * you may find current contact information at www.novell.com
 *
 ****************************************************************************/
{

include "testsuite.ycp";

map READ = $[
    "probe" : $[ "system" : []],
    "target" : $[
	"tmpdir" : "/tmp",
    ]
];

map WRITE = $[];

map EXECUTE = $[
    "target" : $[
	"bash_output": $[ "exit": 0, "stderr": "", "stdout": "" ],
    ],
];
map EXECUTE2 = $[
    "target" : $[
	"bash_output": $[ "exit": 0, "stderr": "", "stdout": "eth0\nppp0\n" ],
    ],
];

TESTSUITE_INIT([READ, WRITE, EXECUTE], nil);

import "Internet";
import "InternetDevices";
import "NetworkInterfaces";

DUMP("Fastest");
READ = $[
    "network" : $[
	"section" : $[ "dsl0":$[], "eth0":$[] ],
	"value" : $[ "dsl0" :$["DEVICE":"eth0"], "eth0":$[] ],
    ],
    "sysconfig": $[ "network": 
		  $["config": 
		   $["NETWORKMANAGER":"yes"]
		   ]
	          ]
];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "dsl0":$[], "eth0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"]= $[ "section" : $[ "eth1":$[], "tr0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "dsl0":$[], "tr0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "dsl0":$[], "ppp0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "ppp0":$[], "tr1":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "ppp0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "ippp0":$[], "tr1":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[ "ippp0":$[], "ppp0":$[] ]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
READ["network"] = $[ "section" : $[]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE], nil);
DUMP(Internet::device);
DUMP(Internet::type);

NetworkInterfaces::initialized = false;
Internet::devices = nil;
READ["network"] = $[ "section" : $[]];
TEST(``(InternetDevices::FindFastest()), [READ, WRITE, EXECUTE2], nil);
DUMP(Internet::device);
DUMP(Internet::type);

}
