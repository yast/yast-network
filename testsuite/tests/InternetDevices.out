Read	.target.tmpdir "/tmp"
Dump	Fastest
Dir	.network.section: ["dsl0", "eth0"]
Dir	.network.value."dsl0": ["DEVICE"]
Read	.network.value."dsl0".DEVICE "eth0"
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Dir	.network.value."eth0": []
Read	.target.stat "/sys/class/net/eth0/type" nil
Read	.target.string "/sys/class/net/eth0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Return	true
Dump	dsl0
Dump	dsl
Dir	.network.section: ["dsl0", "eth0"]
Dir	.network.value."dsl0": []
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Dir	.network.value."eth0": []
Read	.target.stat "/sys/class/net/eth0/type" nil
Read	.target.string "/sys/class/net/eth0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Return	true
Dump	dsl0
Dump	dsl
Dir	.network.section: ["eth1", "tr0"]
Dir	.network.value."eth1": []
Read	.target.stat "/sys/class/net/eth1/type" nil
Read	.target.string "/sys/class/net/eth1/type" nil
Dir	.network.value."tr0": []
Read	.target.stat "/sys/class/net/tr0/type" nil
Read	.target.string "/sys/class/net/tr0/type" nil
Read	.target.stat "/sys/class/net/eth1/type" nil
Read	.target.string "/sys/class/net/eth1/type" nil
Read	.target.stat "/sys/class/net/eth1/type" nil
Read	.target.string "/sys/class/net/eth1/type" nil
Read	.target.stat "/sys/class/net/eth1/type" nil
Read	.target.string "/sys/class/net/eth1/type" nil
Read	.target.stat "/sys/class/net/eth1/type" nil
Read	.target.string "/sys/class/net/eth1/type" nil
Return	true
Dump	eth1
Dump	netcard
Dir	.network.section: ["dsl0", "tr0"]
Dir	.network.value."dsl0": []
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Dir	.network.value."tr0": []
Read	.target.stat "/sys/class/net/tr0/type" nil
Read	.target.string "/sys/class/net/tr0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Return	true
Dump	dsl0
Dump	dsl
Dir	.network.section: ["dsl0", "ppp0"]
Dir	.network.value."dsl0": []
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Dir	.network.value."ppp0": []
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Read	.target.stat "/sys/class/net/dsl0/type" nil
Read	.target.string "/sys/class/net/dsl0/type" nil
Return	true
Dump	dsl0
Dump	dsl
Dir	.network.section: ["ppp0", "tr1"]
Dir	.network.value."ppp0": []
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Dir	.network.value."tr1": []
Read	.target.stat "/sys/class/net/tr1/type" nil
Read	.target.string "/sys/class/net/tr1/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Return	true
Dump	ppp0
Dump	modem
Dir	.network.section: ["ppp0"]
Dir	.network.value."ppp0": []
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Return	true
Dump	ppp0
Dump	modem
Dir	.network.section: ["ippp0", "tr1"]
Dir	.network.value."ippp0": []
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Dir	.network.value."tr1": []
Read	.target.stat "/sys/class/net/tr1/type" nil
Read	.target.string "/sys/class/net/tr1/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Return	true
Dump	ippp0
Dump	isdn
Dir	.network.section: ["ippp0", "ppp0"]
Dir	.network.value."ippp0": []
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Dir	.network.value."ppp0": []
Read	.target.stat "/sys/class/net/ppp0/type" nil
Read	.target.string "/sys/class/net/ppp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Read	.target.stat "/sys/class/net/ippp0/type" nil
Read	.target.string "/sys/class/net/ippp0/type" nil
Return	true
Dump	ippp0
Dump	isdn
Dir	.network.section: []
Execute	.target.bash_output "ip -oneline link list | sed -e 's/^[0-9]*: \\([^:]*\\).*/\\1/' | grep -v 'lo\\|sit0'" $["exit":0, "stderr":"", "stdout":""]
Return	false
Dump	
Dump	
Dir	.network.section: []
Execute	.target.bash_output "ip -oneline link list | sed -e 's/^[0-9]*: \\([^:]*\\).*/\\1/' | grep -v 'lo\\|sit0'" $["exit":0, "stderr":"", "stdout":"eth0\nppp0\n"]
Read	.target.stat "/sys/class/net/eth0/type" nil
Read	.target.string "/sys/class/net/eth0/type" nil
Log	No such device: eth0
Return	true
Dump	eth0
Dump	netcard
